### ISSUE TITLE
eval not parsing correctly

### ISSUE THREAD

**Issue by ElizabethHudnott** [EMGroup/js-eden#174]
Example:

a = 2;
b is a + 2;
c is b + eval(b);

**Comment by knicos** [comment]
in the new parser or both? I was hoping to find a better way of doing this (syntactically) since it is not a function but looks like one and confuses the hell out of me.

**Comment by ElizabethHudnott** [comment]
It's highlighted as a syntax error In the input window.  I don't have the old parser to hand right now, but it worked fine the last time I checked.

If you modify the syntax in the new parser then you will have to change a piece in the maintainer too, since it is re-parsed there.  Basically:

(1) A dependency containing eval could occur inside a proc or an action, in which case the expression shouldn't be substituted with it's value until the proc or action is executed.  Hence it can't be evaluated at parsing time.

(2) However, once the dependency has been stored in the symbol table the value of the expression must remain fixed (i.e. not subject to the usual subscribers/dependencies mechanism).  Also, evaluation happend in isolation then the bit inside the eval in the eden_definition would no longer make any sense (since it refers to the values of observables at definition time, which are not necessarily the same as the current values).

Therefore, at definition time two things happen:

(1) The expressions are evaluated.
(2) The eden_definition gets re-parsed (not fully, but a simple regexp) and altered to substitute the values in place of the eval().  This makes the definition understandable when using the Observable List.  (tkeden also does this rewriting.)

In terms of how it works:

(1) At parse time eval() simply gets replaced by a reference to an index into an object (basically an array with holes in it) which is a global property of the symbol table.
(2) At definition time, the expressions get evaluated and populate those object keys.
(3) When assignment, forget or redefinition happens then the relevant key-value pairs are deleted from the global object.

Relevant files:
Old Parser (jison files)
Old Parser (printEvalIDs helper function in eden.js)
Maintainer

This should be documented somewhere more permanent, since it isn't in Tim's thesis, It's a tkeden feature that was missing that I added later...
